src/Core/Services/CryptoService.cs

```diff
@@ -36,189 +34,268 @@ public class CryptoService : ICryptoService
         {
-            var orgKeys = orgs.ToDictionary(org => org.Id, org => org.Key);
-            _orgKeys = null;
-            await _stateService.SetOrgKeysEncryptedAsync(orgKeys);
+            return await GetUserKeyAsync(userId) != null;
         }
 
-        public async Task<SymmetricCryptoKey> GetKeyAsync(string userId = null)
+        public async Task<bool> HasEncryptedUserKeyAsync(string userId = null)
         {
-            var inMemoryKey = await _stateService.GetKeyDecryptedAsync(userId);
-            if (inMemoryKey != null)
-            {
-                return inMemoryKey;
-            }
-            var key = await _stateService.GetKeyEncryptedAsync(userId);
-            if (key != null)
+            return await _stateService.GetUserKeyMasterKeyAsync(userId) != null;
+        }
+
+        public async Task<UserKey> MakeUserKeyAsync()
+        {
+            return new UserKey(await _cryptoFunctionService.RandomBytesAsync(64));
+        }
+
+        public Task ClearUserKeyAsync(string userId = null)
+        {
+            return _stateService.SetUserKeyAsync(null, userId);
+        }
+
+        public Task SetMasterKeyEncryptedUserKeyAsync(string value, string userId = null)
+        {
+            return _stateService.SetUserKeyMasterKeyAsync(value, userId);
+        }
+
+        public Task SetMasterKeyAsync(MasterKey masterKey, string userId = null)
+        {
+            return _stateService.SetMasterKeyAsync(masterKey, userId);
+        }
+
+        public async Task<MasterKey> GetMasterKeyAsync(string userId = null)
+        {
+            var masterKey = await _stateService.GetMasterKeyAsync(userId);
+            if (masterKey == null)
             {
-                inMemoryKey = new SymmetricCryptoKey(Convert.FromBase64String(key));
-                await _stateService.SetKeyDecryptedAsync(inMemoryKey, userId);
+                // Migration support
+                masterKey = new MasterKey((await _stateService.GetKeyDecryptedAsync(userId)).Key);
+                if (masterKey != null)
+                {
+                    await SetMasterKeyAsync(masterKey, userId);
+                }
             }
-            return inMemoryKey;
+            return masterKey;
+        }
+
+        public Task<MasterKey> MakeMasterKeyAsync(string password, string email, KdfConfig kdfConfig)
+        {
+            return MakeKeyAsync(password, email, kdfConfig, keyBytes => new MasterKey(keyBytes));
+        }
+
+        public Task ClearMasterKeyAsync(string userId = null)
+        {
+            return _stateService.SetMasterKeyAsync(null, userId);
+        }
+
+        public async Task<Tuple<UserKey, EncString>> EncryptUserKeyWithMasterKeyAsync(MasterKey masterKey, UserKey userKey = null)
+        {
+            userKey ??= await GetUserKeyAsync();
+            return await BuildProtectedSymmetricKey(masterKey, userKey.Key, keyBytes => new UserKey(keyBytes));
         }
```

Comment:
Interesting idea, part of my work here was making these kinds of side effects intentional though. Let me know if you'd rather have it inside the method though and I can move it.

Reply:
IMO, if the proposed logic indeed applies to all parts then it should be inside the method just to ease the "work" of the callers and centralize the logic making it easier to fix any potential bugs that may appear regarding the `userKey` parameter, given that only one place needs to be modified (maybe the method name should be updated if this route is taken).
Although I understand the desire of making less black-box some things, I don't think it applies here but if possible I'd like the input from @mpbw2 as well.

Reply:
As long as none of the callers will ever pass in a `UserKey` that _doesn't_ pertain to the current active user, I agree with @fedemkr .  (I'm assuming this is the case since `GetUserKeyAsync()` is called without a `userId`)

Reply:
Sounds good. I wouldn't want to get into a habit of this pattern (if we're missing the `UserKey` or equivalent we might start incorrectly creating new keys and losing data), but it does make sense in this situation.

Reply:
@jlf0dev That's a valid point, so if you foresee cases where it'd be wrong making a new key, by all means leave the code as is, no problem. Better to be safe than to try find a hidden bug in the future because of this.
Also, if the change is done, we can add a `summary` to the method explaining the behavior of it so it's not like a black-box.